implementations/rust/ockam/ockam_command/src/node/create.rs

```
@@ -352,8 +353,46 @@ async fn run_foreground_node(
         }
     }
 
+    // Create a channel for communicating back to the main thread
+    let (tx, mut rx) = tokio::sync::mpsc::channel(2);
+
+    // Register a handler for SIGINT, SIGTERM, SIGHUP
+    let tx_clone = tx.clone();
+    let mut opts_clone = opts.clone();
+    ctrlc::set_handler(move || {
+        tx_clone.blocking_send(()).unwrap();
+        opts_clone
+            .shell
+            .write_line(format!("{} Ctrl+C signal received", "!".light_yellow()).as_str())
+            .unwrap();
+    })
+    .expect("Error setting Ctrl+C handler");
+
+    // Spawn a thread to monitor STDIN for EOF
     if cmd.exit_on_eof {
-        stop_node_on_eof(&mut ctx, &opts, &node_name).await?;
+        let tx_clone = tx.clone();
+        let mut opts_clone = opts.clone();
+        std::thread::spawn(move || {
+            let mut buffer = Vec::new();
+            let mut handle = io::stdin().lock();
+            handle
+                .read_to_end(&mut buffer)
+                .expect("Error reading from stdin");
+            opts_clone
+                .shell
+                .write_line(format!("{} EOF received", "!".light_yellow()).as_str())
+                .unwrap();
+            let _ = tx_clone.blocking_send(());
```

Let's send the message first and then log the message. In this context, we can ignore the error as it can't be recovered and we don't care if we can't write to stderr:

```rust
let _ = tx_clone.blocking_send(());
let _ = opts_clone.shell.write(...);
```